File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int _row = matrix.size() - 1;
        
        for (size_t i = 0u; i < matrix.size() - 1; i++){
            if (matrix[i][0u] <= target && matrix[i + 1][0u] > target){
                _row = i;
            }
        }

        if (matrix[_row].back() < target || matrix[_row].front() > target){
            return false;
        }
        
        vector<int>& search_line = matrix[_row];
        
        int left = 0u;
        int right = search_line.size() - 1;
        
        //binarySearch
        while (left <= right) {
            int mid = (left + right) / 2;

            if (search_line[mid] == target) {
                return true;
            }

            if (search_line[mid] > target) {
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
	    }   
        
        return false;
    }
};

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int left = 0;
        int right = matrix.length - 1;
        int mid = -1;
        int end_Indx = matrix[0].length - 1;

        if (matrix[right][end_Indx] < target){
            return false;
        }

        while(left <= right){
            mid = (left + right) / 2;

            if (matrix[mid][0] <= target && matrix[mid][end_Indx] >= target) {
                break;
            }

            if (matrix[mid][0] > target){
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }

        }

        if (left > right){
            return false;
        }

        int row = mid;

        int _left = 0;
        int _right = matrix[0].length - 1;
        int _mid = -1;

        while (_left <= _right){
            _mid = (_left + _right) / 2;

            if (matrix[row][_mid] == target){
                return true;
            }

            if (matrix[row][_mid] < target){
                _left = _mid + 1;
            }
            else {
                _right = _mid - 1;
            }
        }

        return false;
    }
}